### PR TITLE
feat(Dgraph): Online restores allows to restore a specific backup.

### PR DESCRIPTION
**Pull request by martinmr**
Currently online restores will restore the entire backup series. This PR allows to
restore the backup series up to a specific backup.

Fixes DGRAPH-1648

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6411)

 
 
 

Docs Preview: [Dgraph Preview](https://dgraph-db9a5f39f7-93443.surge.sh)

### CODE REVIEW

**Comment by martinmr** [comment]
Disregard the changes to the docker-compose files and the makefiles. They are a workaround to an issue I'm having (my workstation has a different version of glibc than the docker image). I will revert these changes before I merge this PR.

File: protos/pb.proto [hunk 1]
```diff
@@ -287,6 +287,8 @@ message RestoreRequest {
 	string vault_path = 13;
 	string vault_field = 14;
 	string vault_format = 15;
+
+  uint64 backup_num = 16;
```
Comment (parasssh) [inline protos/pb.proto:291]: nit: formatting is messed.

**Review by parasssh — COMMENTED** [review]
Took a stab. Lets do synchronous review so I can understand better.

**Review by parasssh — APPROVED** [review]
:lgtm: minor comments. Feel free to resolve. 

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-:-MHCJBXnEwlQdpFozj9R:b-sc8qh9)* status: 0 of 13 files reviewed, 5 unresolved discussions (waiting on @manishrjain, @martinmr, @MichaelJCompton, @pawanrawal, and @vvbalaji-dgraph)

___
*[dgraph/Makefile, line 46 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCHo8D2m46U_Va4zk7:-MHCHo8EBGGR0jqQjbe1:bauo7li) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/dgraph/Makefile#L46)):*
> ```Makefile
> lastCommitTime  = github.com/dgraph-io/dgraph/x.lastCommitTime
> 
> BUILD_FLAGS   ?= -ldflags '-X ${lastCommitSHA}=${BUILD} -X "${lastCommitTime}=${BUILD_DATE}" -X "${dgraphVersion}=${BUILD_VERSION}" -X "${dgraphCodename}=${BUILD_CODENAME}${MODIFIED}" -X ${gitBranch}=${BUILD_BRANCH}' -tags netgo
> ```

what is netgo tag for?
___
*[worker/backup_handler.go, line 66 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCIIBD0ChZxhSTTc-b:-MHCIIBE3gXTrfDAX_vs:b3qvm1h) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/worker/backup_handler.go#L66)):*
> ```Go
> 	// and backup number at the specified location. If backupNum is set to zero,
> 	// all the manifests for the backup series will be returned.
> 	GetManifests(*url.URL, string, uint64) ([]*Manifest, error)
> ```

Maybe clarify comment that given a backupNum, it returns manifests from 0 to thru the backupNum. Not clear as it is. 
___
*[worker/backup_handler.go, line 318 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCIkcFEpQtCYIaZaK8:-MHCIkcFEpQtCYIaZaK9:b-m8s2ir) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/worker/backup_handler.go#L318)):*
> ```Go
> 
> 	if backupNum > 0 {
> 		if len(manifests) < int(backupNum) {
> ```

are we off by 1? e.g. if backupNum = 1, i expect 2 manifests at 0 and 1, no?

**Review by martinmr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6411)* status: 0 of 11 files reviewed, 3 unresolved discussions (waiting on @manishrjain, @MichaelJCompton, @parasssh, @pawanrawal, and @vvbalaji-dgraph)

___
*[dgraph/Makefile, line 46 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCHo8D2m46U_Va4zk7:-MHD0x3ACKpCFxQw8sUB:b-g3iz7d) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/dgraph/Makefile#L46)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

what is netgo tag for?
</blockquote></details>

Ignore that. It's just because I couldn't build Dgraph on my machine. I'll revert that.

But the tag is to build the net library without using the c standard library.
___
*[systest/online-restore/online_restore_test.go, line 178 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCAX6ND82Jch00zcXd:-MHD1N-pE0DBJpU-yJel:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/systest/online-restore/online_restore_test.go#L178)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

> ```
> t.Logf(string(resp.GetJson()))
> 			t.Logf(string(bodies[1]))
> ```
revert this
</blockquote></details>

Done.
___
*[worker/backup_handler.go, line 63 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCA2mWDU8nFORxt7cO:-MHD1V27AvQKsWIxJoIH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/worker/backup_handler.go#L63)):*
<details><summary><i>Previously, martinmr (Martin Martinez Rivera) wrote…</i></summary><blockquote>

> ```
>  GetManfiest
> ```
typo
</blockquote></details>

Done.
___
*[worker/backup_handler.go, line 66 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCIIBD0ChZxhSTTc-b:-MHD1dxcAdX8IF1g00AW:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/worker/backup_handler.go#L66)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

Maybe clarify comment that given a backupNum, it returns manifests from 0 to thru the backupNum. Not clear as it is. 
</blockquote></details>

Done.
___
*[worker/backup_handler.go, line 318 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6411#-MHCIkcFEpQtCYIaZaK8:-MHD1evQCAPDkfEAtvG7:b-tadz9i) ([raw file](https://github.com/dgraph-io/dgraph/blob/6166c408c49586c368fe6f505e67088b039e1023/worker/backup_handler.go#L318)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

are we off by 1? e.g. if backupNum = 1, i expect 2 manifests at 0 and 1, no?
</blockquote></details>

There's no backup with backupNum 0. We start the count at 1.

**Pull request merged**